[API] Patchable Budgets

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19141
## Dependencies

Requires merge of:
- [x] https://github.com/opf/openproject/pull/2708
- [x] https://github.com/opf/openproject/pull/2740
## TODO
- [x] document budget endpoints and links
- [x] add budget endpoints
- [x] link budget in WP
- [x] add budget to schema
- [x] add budget to payload (actually making it PATCH-able)
- [x] ~~adapt front end after renaming cost_object ("budget")~~
### Integration tests
- [ ] GET work package
- [x] POST work package form
- [x] PATCH work package
- [x] GET budget
- [x] GET budgets by project
## TODOs to maintain consistency
- [x] document changed behaviour of spent time on WP
- [x] reflect changed behaviour of spent time in schema

NobodysNightmare committed:
ignore RubyMine files

NobodysNightmare committed:
rename cost_objects to budgets in API

NobodysNightmare committed:
let budget representer use Single as base

NobodysNightmare committed:
add documentation for costs API endpoints

NobodysNightmare committed:
add real endpoint for budgets

- use minimal representer for now
- solution to add the endpoint is only mediocre

NobodysNightmare committed:
add API endpoint for budgets by project

NobodysNightmare committed:
include schema for budget

NobodysNightmare committed:
modify visibility of spentTime in schema

NobodysNightmare committed:
document changes in spentTime behaviour

NobodysNightmare committed:
fix cost_object validation

- would not work for invalid ids
- use "inclusion" message instead of generic "invalid" message

NobodysNightmare committed:
allow to actually change the budget via API

NobodysNightmare committed:
render budget if the user may :view_budgets

NobodysNightmare committed:
style

NobodysNightmare committed:
copy shared example from core

not available in a shared form anymore...
I would be happier to share it among core and plugins, but that would assume that all shared users define a let(:embedded) that could be overridden...

NobodysNightmare committed:
change set of disabled specs

- one new spec that is failing
- all other specs now run without problems

NobodysNightmare commented:
Note to be safe:

I reduced the amount of properties available in the representer for cost objects. I don't think that the Angular client relied on anything beyond subject and id, but I could be wrong.

ulferts commented on lib/open_project/costs/patches/work_package_patch.rb:
```diff
@@ -94,10 +94,14 @@ def update_cost_entries(work_packages, action)
-      if cost_object && cost_object.changed?
-        unless (cost_object.blank? || project.cost_object_ids.include?(cost_object.id))
-          errors.add :cost_object, :invalid
+      if changed.include? 'cost_object_id'
```
`cost_object_id_changed?` is easier to read

ulferts commented:
@NobodysNightmare the budget is no longer displayed, because of the renaming from cost_object to budget. You want to give it a try in the front end?

ulferts commented:
@NobodysNightmare I tried unsuccessfully to patch the work package to have a new budget. And there is no integration test that let's me assume that this should be working. On the other hand the PR is named "Patchable Budgets"...

NobodysNightmare committed:
use more readable "changed" test

NobodysNightmare commented:
I chose not to do an integrational test for `GET` work package.
Because we already have a spec for the representer and I don't see which extra benefit the inegration test would offer.

However, I will create the other integration tests. They seem useful.

NobodysNightmare committed:
add integration tests for the budget

NobodysNightmare commented:
> You want to give it a try in the front end?

I did, but I am apparently too stupid...

~~@0xF013 this is your turn again, what do I have to do here? It is not `_embedded/costObject` anymore, but `_embedded/budget`. Can't be too hard, right?~~

edit: see below

NobodysNightmare commented:
I changed the link back to `costObject`, because the waqy we generate the field name for errors in our form sucks and I am either forced to refactor there or simply stay with `costObject` for the time being...

Still able to change that in the future, right?

NobodysNightmare committed:
rename the link to the budget of a work package

due to a shortcoming of the way we generate the errors in a form this
is neccessary for now, to keep the form consistent

NobodysNightmare commented:
Note that there is a new (small) dependency PR in the core.

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/dev' into feature/budgets_api

Conflicts:
	lib/open_project/costs/engine.rb
